crates/services/executor/src/executor.rs

```diff
@@ -1430,33 +1436,61 @@ where
             .collect();
         let ready_tx = checked_tx.into_ready(gas_price, gas_costs, fee_params)?;
 
-        let vm_result: StateTransition<_> = vm
+        let mut vm = Interpreter::with_storage(
+            vm_db,
+            InterpreterParams::new(gas_price, &self.consensus_params),
+        );
+
+        // We get memory from the pool and are careful to return it there when done
+        *vm.memory_mut() = self.vm_pool.get_new();
+
+        // This returns the VM memory to the pool to before returning
+        macro_rules! recycling_try {
+            ($expr:expr) => {
+                match $expr {
+                    Ok(val) => val,
+                    Err(e) => {
+                        self.vm_pool.recycle(mem::take(vm.memory_mut()));
+                        return Err(e.into())
+                    }
+                }
+            };
+        }
+
+        let vm_result: StateTransition<_> = recycling_try!(vm
             .transact(ready_tx)
             .map_err(|error| ExecutorError::VmExecution {
                 error: error.to_string(),
                 transaction_id: tx_id,
-            })?
-            .into();
+            }))
+        .into();
         let reverted = vm_result.should_revert();
 
         let (state, mut tx, receipts): (_, Tx, _) = vm_result.into_inner();
         #[cfg(debug_assertions)]
         {
-            tx.precompute(&self.consensus_params.chain_id())?;
+            recycling_try!(tx.precompute(&self.consensus_params.chain_id()));
             debug_assert_eq!(tx.id(&self.consensus_params.chain_id()), tx_id);
         }
 
-        Self::update_input_used_gas(predicate_gas_used, tx_id, &mut tx)?;
+        recycling_try!(Self::update_input_used_gas(
+            predicate_gas_used,
+            tx_id,
+            &mut tx
+        ));
 
         // We always need to update inputs with storage state before execution,
         // because VM zeroes malleable fields during the execution.
-        self.compute_inputs(tx.inputs_mut(), storage_tx)?;
+        recycling_try!(self.compute_inputs(tx.inputs_mut(), storage_tx));
 
         // only commit state changes if execution was a success
         if !reverted {
             self.log_backtrace(&vm, &receipts);
+            self.vm_pool.recycle(mem::take(vm.memory_mut()));
             let changes = sub_block_db_commit.into_changes();
             storage_tx.commit_changes(changes)?;
+        } else {
+            self.vm_pool.recycle(mem::take(vm.memory_mut()));
```

Comment:
Since we are always recycling the memory, why do we even need the `recycling_try` macro?

Reply:
Since we need to recycle just before returning early. The continuing branch still needs the vm.

Reply:
I made it quite a bit prettier on the fuel-vm side, now it just auto-recycles on drop.